---
title: "Isoformic - Recategorize"
date: "2023-04-07"
output:
  rmarkdown::html_document:
    highlight: tango
    toc: true
    toc_float: true
author:
  - name: Izabela Mamede Conceição
    affiliation: Laboratory of Genetics Biochemistry, Universidade Federal de Minas Gerais, Belo Horizonte, Brazil
  - name: Lucio Rezende Queiroz
    affiliation: Department of Pathology and Laboratory Medicine, Weill Cornell Medicine, New York, USA
vignette: |
  %\VignetteIndexEntry{intro-isoformic}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: sentence
---

# Re-categorizing transcripts based on a reference

The last issue we try to explore with isoformic is a re-classification of transcripts based on their modifications compared to the *canonical transcript*, the one that is used as the "gene" reference, and the quantification of this modifications in a gene-group basis. 

The biggest problem in this type of analysis is determine which would be the "reference" transcript of a gene. This reference biologically will vary a lot from tissue to tissue but to do this type of quantification needs to be determined. To bypass this discussion here we decided to use the MANE (Matched Annotation from NCBI and EMBL-EBI) list as the reference, since it  combines both NCBI and EMBL annotations and possess most (around 95%) of all the human genes.

For this we determined four categories: alternative transcription start site (aTSS), alternative transcription end site (aTES), exon skipping an intron retention. All of these are based comparing each transcript information (extracted from a gencode .gff) with the MANE-transcript information. Transcripts which have the same start, the same end and the same exon number when compared to the MANE reference will be categorized as "Others".

```{r}
knitr::include_graphics("https://i.imgur.com/i5zcwVW.png")
```

First then we do the formatting of the MANE and the GENCODE gff using the format_gff

```{r}
formatgff <- function(MANE_gtf_pathway, gencode_gff_pathway) {
  suppressWarnings({
    suppressMessages({
      MANE_GRCh38_v1_2_ensembl_genomic <- 
  read_table(MANE_gtf_pathway, col_names = FALSE)
transcripts<-MANE_GRCh38_v1_2_ensembl_genomic$X12
transcripts_MANE<- transcripts[stringr::str_detect(transcripts, "ENST")]
transcripts_MANE_f<-gsub('"', '', transcripts_MANE)
transcripts_MANE_f<-unique(gsub(';', '', transcripts_MANE_f))
mane_transcripts<-tx_to_gene %>% filter(transcript_id %in% transcripts_MANE_f)
gencode_v34_annotation <- read_delim(gencode_gff_pathway, 
                                     delim = "\t", escape_double = FALSE, 
                                     col_names = FALSE, trim_ws = TRUE, skip = 7)
gencode_v34_annotation_tomod<-gencode_v34_annotation
gencode_v34_annotation_tomod[1:12] <-
  gencode_v34_annotation_tomod$X9 %>% str_split_fixed(';', 12)
gff_v34<-cbind(gencode_v34_annotation, gencode_v34_annotation_tomod)
names(gff_v34) <- c("chr", "HAVANA", "type", "left_pos", "right_pos", "exclude_1",
                    "strand", "exclude_2", "etc", "ID", "V1", "V2", "V3", "V4",
                    "V5", "V6", "V7", "V8", "V9", "V10", "V11")
gff_v34_filt <-gff_v34 %>% select(3:5, 7, 10)
gff_v34_filt$ID <- str_sub(gff_v34_filt$ID, start = 4)
gff_v34_filt$ID <- str_replace(gff_v34_filt$ID, "exon:", "")
df_split <- separate(gff_v34_filt, ID, into = c("tx_id", "exon_count"), sep = ":")
df_split_gene<-df_split %>% filter(type =="gene")
df_split_notgene<-df_split %>% filter(type !="gene")
df_split_gene_dic<-df_split_gene %>% left_join(tx_to_gene, by =c("tx_id"="gene_id"))
df_split_gene_dic<-df_split_gene_dic %>% select(1:6, 9)
df_split_gene_dic <- distinct(df_split_gene_dic)
df_split_notgene_dic<-df_split_notgene %>% left_join(tx_to_gene, by =c("tx_id"="transcript_id"))
df_split_notgene_dic<-df_split_notgene_dic %>% select(1:6, 9)
df_with_all<-rbind(df_split_gene_dic, df_split_notgene_dic)
df_transcripts_only<-df_with_all %>% filter(type == "transcript")
MANE_dic<-mane_transcripts %>%
  left_join(df_transcripts_only, by =c("transcript_id"="tx_id"))
mane_filt<-MANE_dic %>% select(gene_name.x, left_pos, right_pos)
names(mane_filt) <- c("gene_name" ,"left_pos"   , "right_pos" )
result <-list(mane_filt = mane_filt, df_all = df_with_all)
return(result)
    })
  })
}

result_format <- format_gff(MANE_gtf_pathway =
                              "Ref/MANE.GRCh38.v1.2.ensembl_genomic.gtf",
                            gencode_gff_pathway =
                              "Ref/gencode.v34.annotation.gff3")
```

This function outputs a list with two tables the formatted gff information and the formatted MANE information

```{r}
head(result_format$mane_filt)
```

With those two tables we can then run the isoformic_categorize function which takes a vector of genes as input and retirns the classification of all the transcripts of that gene. if the gene you are inputting is *not* a MANE gene the function will *not* work since there will be no reference to compare.

It is possible to run this function in a vector using all the MANE genes but it takes quite a while so here we will use a small vector of genes to do it as a reference

```{r}
isoformic_categorize <- function(gene_vector, df_with_all, mane_filt) {
  table_categorized <- list()
  
  for (selected_gene in gene_vector) {
    gtf_df <- df_with_all %>% filter(gene_name == selected_gene)
    gtf_df$exon_count <- as.numeric(gtf_df$exon_count)
    canon_table <- gtf_df %>% filter(type == "gene")
    mane_default <- mane_filt %>% filter(gene_name == selected_gene)
    default_left <- mane_default$left_pos
    default_right <- mane_default$right_pos
    canon_transcript_table <- gtf_df %>% filter(left_pos == default_left &
                                                  right_pos == default_right &
                                                  type == "transcript")
    default_transcript <- canon_transcript_table$tx_id
    default_transcript_tab <- gtf_df %>% filter(tx_id == default_transcript)
    
    default_exon_count <- max(default_transcript_tab$exon_count, na.rm = TRUE)
    
    exon_counts <- gtf_df %>%
      select(1, 4, 5, 6, 7) %>%
      filter(type == "exon") %>%
      group_by(tx_id) %>%
      summarise(max(exon_count))
    names(exon_counts) <- c("tx_id", "max_exon")
    
    gtf_df_counts <- gtf_df %>% 
      left_join(exon_counts, by = c("tx_id" = "tx_id"))
    
    transcripts_with_types <- gtf_df_counts %>%
      filter(type == "transcript") %>%
      mutate(category = case_when(
        (strand == "+" & left_pos != default_left & right_pos == default_right &
           max_exon == default_exon_count) ~ "aTSS",
        (strand == "+" & left_pos == default_left & right_pos != default_right &
           max_exon == default_exon_count) ~ "aTES",
        (strand == "+" & left_pos == default_left & right_pos == default_right &
           max_exon > default_exon_count) ~ "intron_retained",
        (strand == "+" & left_pos == default_left & right_pos == default_right &
           max_exon < default_exon_count) ~ "exon_skipping",
        (strand == "+" & left_pos != default_left & right_pos != default_right &
           max_exon < default_exon_count) ~ "aTSS aTES exon_skipping",
        (strand == "+" & left_pos != default_left & right_pos != default_right &
           max_exon > default_exon_count) ~ "aTSS aTES intron_retained",
        (strand == "+" & left_pos != default_left & right_pos != default_right &
           max_exon == default_exon_count) ~ "aTSS aTES",
        (strand == "+" & left_pos != default_left & right_pos == default_right &
           max_exon < default_exon_count) ~ "aTSS exon_skipping",
        (strand == "+" & left_pos != default_left & right_pos == default_right &
           max_exon > default_exon_count) ~ "aTSS intron_retained",
        (strand == "+" & left_pos == default_left & right_pos != default_right &
           max_exon > default_exon_count) ~ "aTES exon_skipping",
        (strand == "+" & left_pos == default_left & right_pos != default_right &
           max_exon < default_exon_count) ~ "aTES intron_retained",
        
        (strand == "-" & left_pos != default_left & right_pos == default_right &
           max_exon == default_exon_count) ~ "aTES",
        (strand == "-" & left_pos == default_left & right_pos != default_right &
           max_exon == default_exon_count) ~ "aTSS",
        (strand == "-" & left_pos == default_left & right_pos == default_right &
           max_exon > default_exon_count) ~ "intron_retained",
        (strand == "-" & left_pos == default_left & right_pos == default_right &
           max_exon < default_exon_count) ~ "exon_skipping",
        (strand == "-" & left_pos != default_left & right_pos != default_right &
           max_exon < default_exon_count) ~ "aTSS aTES exon_skipping",
        (strand == "-" & left_pos != default_left & right_pos != default_right &
           max_exon > default_exon_count) ~ "aTSS aTES intron_retained",
        (strand == "-" & left_pos != default_left & right_pos != default_right &
           max_exon == default_exon_count) ~ "aTSS aTES",
        (strand == "-" & left_pos != default_left & right_pos == default_right &
           max_exon < default_exon_count) ~ "aTES exon_skipping",
        (strand == "-" & left_pos != default_left & right_pos == default_right &
           max_exon > default_exon_count) ~ "aTSS intron_retained",
        (strand == "-" & left_pos == default_left & right_pos != default_right &
           max_exon > default_exon_count) ~ "aTSS exon_skipping",
        (strand == "-" & left_pos == default_left & right_pos != default_right &
           max_exon < default_exon_count) ~ "aTSS intron_retained",
        (strand == "-" & left_pos == default_left & right_pos == default_right &
           max_exon == default_exon_count) ~ "Canonic",
        TRUE ~ "Others"
      ))
    
    table_categorized[[selected_gene]] <- transcripts_with_types %>% select(5, 7, 9)
  }
  
  combined_table_df <- do.call(rbind, table_categorized)
  rownames(combined_table_df) <- NULL
  return(combined_table_df)
}

```

run

```{r}
vector_with_genes <- c("PAPPA2", "CYP2J2", "FLT1", "PHLDB2")

genes_recategorized <-isoformic_categorize(gene_vector = vector_with_genes,
                            mane_filt =  result_format$mane_filt, 
                            df_with_all =  result_format$df_all)

```

We can use the data-frame produced for the profile-plot to visualize the counts of the different types of transcripts between case and control and also inside the pathways that were returned on enrichment. A good way to visualize this would be any graph that can compute well proportions, a barplot would do, but to be able to visualize groups and subgroups we opted for a treemap type graph.

First get from the count data the genes which transcripts we just recategorized and then re-plot.

```{r}
tx_to_tx<-tx_to_gene %>% select(transcript_id, transcript_name)

genes_recategorized_dic<-genes_recategorized %>% left_join(tx_to_tx, by = c("tx_id"="transcript_id"))

genes_recategorized_dic <- genes_recategorized_dic %>% left_join(profile_data_df, by =c("transcript_name"="genename"))
```

We can plot this proportions per gene and even for all the genes in a certain transcript group, first a color scheme for the groups

```{r}
categories_vector <-c("aTSS", "aTES", "intron_retained", "exon_skipping", "aTSS aTES exon_skipping",
  "aTSS aTES intron_retained", "aTSS aTES", "aTSS exon_skipping", "aTSS intron_retained", "aTES exon_skipping", "aTES intron_retained", "Canonic", "Others")

cat_colors <- c("#b15928", "#33a02c", "grey","#a6cee3", "#ffff99", "#1f78b4", "#ff7f00", "#fdbf6f", "purple", "#e31a1c", "#cab2d6","#b2df8a","#fb9a99")

names(cat_colors) <- categories_vector

```

We can plot these genes separated with their transcripts using treemapfy

For treatment TPM
```{r}
genes_recategorized_dic %>%
  drop_na(condition) %>%
  #filter(condition == "treatment") %>%
ggplot(aes(area = mean_TPM, fill = category, subgroup = condition)) +
  geom_treemap()+
  geom_treemap_text(aes(label = transcript_name), grow = FALSE, color = "grey8")+
   geom_treemap_subgroup_border()+
  geom_treemap_subgroup_text(color = "white",size = 10)+
  facet_wrap(~gene_name)+
  theme_bw()+
  scale_fill_manual(values = cat_colors)+
  ggtitle("Recategorized transcripts TPM between case and control")

```

It is possible to also use this information with the pathway results. Here an example using the df_enrichment output from the functional transcript enrichment for the Angiogenesis pathway genes.

First we isoformic recategorize those transcripts
```{r}
enrichment_angiogenesis_productive<-enrichment_df %>% filter(pathway == "HALLMARK_ANGIOGENESIS" & experiment == "protein_coding")

genes_angiogenesis_prod<-mane_filt %>% filter(gene_name %in% unlist(enrichment_angiogenesis_productive$leadingEdge)) 

angiogenesis_recategorized <-isoformic_categorize(gene_vector = genes_angiogenesis_prod$gene_name,
                            mane_filt =  result_format$mane_filt, 
                            df_with_all =  result_format$df_all)
```

Then we do the joining to get the TPM information from the profile table

```{r}
angiogenesis_recategorized_dic<-angiogenesis_recategorized %>% left_join(tx_to_tx, by = c("tx_id"="transcript_id"))

angiogenesis_recategorized_dic <- angiogenesis_recategorized_dic %>% left_join(profile_data_df, by =c("transcript_name"="genename"))

```

And the plotting

```{r}
angiogenesis_recategorized_dic %>%
  drop_na(condition) %>%
  filter(mean_TPM > 20) %>%
ggplot(aes(area = mean_TPM, fill = category, subgroup = condition)) +
  geom_treemap()+
  geom_treemap_text(aes(label = transcript_name), grow = FALSE, color = "grey8")+
   geom_treemap_subgroup_border()+
  geom_treemap_subgroup_text(color = "white",size = 10)+
  facet_wrap(~gene_name)+
  theme_bw()+
  scale_fill_manual(values = cat_colors)+
  ggtitle("Angiogenesis transcripts TPM between case and control")

```

# References
[ensembl-ref]: <https://ensembl.org/>
[gencode-ref]: <https://www.gencodegenes.org/human/>
[fgsea-ref]: <https://bioconductor.org/packages/release/bioc/html/fgsea.html>
[swish-ref]: <https://bioconductor.org/packages/release/bioc/vignettes/fishpond/inst/doc/swish.html>
[pe-paper-ref]: <https://www.ahajournals.org/doi/10.1161/HYPERTENSIONAHA.120.14756>
[gmt-format-ref]: <https://software.broadinstitute.org/cancer/software/gsea/wiki/index.php/Data_formats>
[mane-ref]: <https://www.ncbi.nlm.nih.gov/refseq/MANE/>




